source/mqtt/MqttConnectionCore.cpp

```diff
@@ -0,0 +1,891 @@
+                            // TODO Use emplace_back.
+                            topics.emplace_back(
+                                reinterpret_cast<char *>(subscription->topic.ptr), subscription->topic.len);
+                            // TODO Is only the last one needed?
+                            qos = subscription->qos;
+                        }
+
+                        callbackData->onSubAck(*connection, packetId, topics, qos, errorCode);
+                    }
+                }
+
+                // Clean up previously allocated resources.
+                if (callbackData->topic != nullptr)
+                {
+                    aws_mem_release(
+                        callbackData->allocator, reinterpret_cast<void *>(const_cast<char *>(callbackData->topic)));
+                }
+                Crt::Delete(callbackData, callbackData->allocator);
+            }
+
+            void MqttConnectionCore::createUnderlyingConnection(aws_mqtt_client *client)
+            {
+                m_underlyingConnection = aws_mqtt_client_connection_new(client);
+            }
+
+            void MqttConnectionCore::createUnderlyingConnection(aws_mqtt5_client *mqtt5Client)
+            {
+                m_underlyingConnection = aws_mqtt_client_connection_new_from_mqtt5_client(mqtt5Client);
+            }
+
+            void MqttConnectionCore::connectionInit()
+            {
+                if (m_underlyingConnection != nullptr)
+                {
+                    aws_mqtt_client_connection_set_connection_result_handlers(
+                        m_underlyingConnection,
+                        MqttConnectionCore::s_onConnectionSuccess,
+                        this,
+                        MqttConnectionCore::s_onConnectionFailure,
+                        this);
+
+                    aws_mqtt_client_connection_set_connection_interruption_handlers(
+                        m_underlyingConnection,
+                        MqttConnectionCore::s_onConnectionInterrupted,
+                        this,
+                        MqttConnectionCore::s_onConnectionResumed,
+                        this);
+
+                    aws_mqtt_client_connection_set_connection_closed_handler(
+                        m_underlyingConnection, MqttConnectionCore::s_onConnectionClosed, this);
+
+                    aws_mqtt_client_connection_set_connection_termination_handler(
+                        m_underlyingConnection, MqttConnectionCore::s_onConnectionTermination, this);
+
+                    m_isConnectionAlive = true;
+                }
+                else
+                {
+                    AWS_LOGF_DEBUG(AWS_LS_MQTT_CLIENT, "Failed to initialize Mqtt Connection");
+                }
+            }
+
+            std::shared_ptr<MqttConnection> MqttConnectionCore::obtainConnectionInstance()
+            {
+                // FIXME Okay, this actually doesn't work. There is a chance that the MqttConnection dtor started
```

Comment:
This sounds alarming.  It seems like we're doing things pretty differently than the 5 client and I don't see why.  I'm not sure I see a path to correctness with this approach.

Reply:
I checked once again, and the case I described in this comment is not possible after all. If the `MqttConnection` destructor started its execution, that means that the counter in `std::shared_ptr` reached 0. So, `std::weak_ptr::lock()` won't return object.

Reply:
Ok, I think I follow now.  I am still skeptical of m_isConnectionAlive though.

Reply:
Actually, with the current implementation, `m_isConnectionAlive` is not required anymore: `std::weak_ptr` handles the liveliness of the `MqttConnection` object. Moreover, the mutex is not needed anymore as well.

Reply:
I added a comment describing how it works.